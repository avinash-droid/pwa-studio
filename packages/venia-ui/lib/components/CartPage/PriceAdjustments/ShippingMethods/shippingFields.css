.country {
    grid-area: country;
}

.state {
    grid-area: state;
}

.zip {
    grid-area: zip;
}
