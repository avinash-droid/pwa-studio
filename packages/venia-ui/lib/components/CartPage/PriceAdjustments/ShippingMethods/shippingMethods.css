.root {
    display: grid;
    gap: 1rem 2rem;
    grid-template-columns: 2fr 1fr;
    grid-template-areas:
        'country country'
        'state zip'
        'methods methods';
    padding-right: 6rem;
}

.country {
    grid-area: country;
}

.prompt {
    font-weight: bold;
    padding: 0.5em 0em;
}

.shipping_methods {
    grid-area: methods;
}

.shipping_methods_hidden {
    composes: shipping_methods;
    display: none;
}

.state {
    grid-area: state;
}

.zip {
    grid-area: zip;
}

/*
 * Mobile-specific styles.
 */

@media (max-width: 640px) {
    .root {
        /* Only one column in mobile view. */
        grid-template-columns: 100%;
        grid-template-areas:
            'country'
            'state'
            'zip'
            'methods';
        padding-right: unset;
    }
}
