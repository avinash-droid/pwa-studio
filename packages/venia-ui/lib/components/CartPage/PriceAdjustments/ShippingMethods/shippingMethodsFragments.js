import gql from 'graphql-tag';

export const ShippingMethodsFragment = gql`
    fragment ShippingMethodsFragment on Cart {
        id
        shipping_addresses {
            available_shipping_methods {
                amount {
                    currency
                    value
                }
                available
                carrier_code
                carrier_title
                method_code
                method_title
            }
        }
    }
`;
