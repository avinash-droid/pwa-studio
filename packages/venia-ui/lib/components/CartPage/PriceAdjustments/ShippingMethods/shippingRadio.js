import React, { Fragment } from 'react';
import { Price } from '@magento/peregrine';

const ShippingRadio = props => {
    const priceElement = props.price ? (
        <Price value={props.price} currencyCode={props.currency} />
    ) : (
        <span>FREE</span>
    );

    return (
        <Fragment>
            <span>{props.name}</span>
            <div>{priceElement}</div>
        </Fragment>
    );
};

export default ShippingRadio;
