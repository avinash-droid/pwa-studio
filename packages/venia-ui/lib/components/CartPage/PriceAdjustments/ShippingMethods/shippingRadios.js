import React from 'react';

import RadioGroup from '../../../RadioGroup';

import { mergeClasses } from '../../../../classify';
import defaultClasses from './shippingRadios.css';
import ShippingRadio from './shippingRadio';

const ShippingRadios = props => {
    const { shippingMethods } = props;
    const radioComponents = shippingMethods.map(shippingMethod => ({
        key: shippingMethod.method_code,
        label: (
            <ShippingRadio
                currency={shippingMethod.amount.currency}
                name={shippingMethod.method_title}
                price={shippingMethod.amount.value}
            />
        ),
        value: {
            carrier_code: shippingMethod.carrier_code,
            method_code: shippingMethod.method_code
        }
    }));
    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <RadioGroup
            classes={{
                radio: classes.radio,
                radioLabel: classes.radio_contents,
                root: classes.root
            }}
            field="method"
            items={radioComponents}
        />
    );
};

export default ShippingRadios;
