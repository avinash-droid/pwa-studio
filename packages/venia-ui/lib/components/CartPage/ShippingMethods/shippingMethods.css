.root {
    display: grid;
    gap: 1rem;
    grid-template-columns: 75% auto;
    grid-template-areas:
        'country country'
        'state zip'
        'methods methods';
    margin: 0.5em;
}

.country {
    grid-area: country;
}

.prompt {
    font-weight: bold;
    padding: 0.5em 0em;
}

.shipping_methods {
    grid-area: methods;
}

.state {
    grid-area: state;
}

.zip {
    grid-area: zip;
}

/*
 * Mobile-specific styles.
 */

 @media (max-width: 640px) {
    .root {
        /* Only one column in mobile view. */
        grid-template-columns: 100%;
        grid-template-areas:
            'country'
            'state'
            'zip'
            'methods';
    }
}