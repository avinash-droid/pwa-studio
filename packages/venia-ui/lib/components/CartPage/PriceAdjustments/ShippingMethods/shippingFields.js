import React, { Fragment } from 'react';
import gql from 'graphql-tag';
import { func, shape, string } from 'prop-types';
import { useShippingFields } from '@magento/peregrine/lib/talons/CartPage/PriceAdjustments/ShippingMethods/useShippingFields';

import { mergeClasses } from '../../../../classify';
import Field from '../../../Field';
import Select from '../../../Select';
import TextInput from '../../../TextInput';
import defaultClasses from './shippingFields.css';
import { ShippingMethodsFragment } from './shippingMethodsFragments';

const ShippingFields = props => {
    const { selectedShippingFields, setIsFetchingMethods } = props;
    const {
        countries,
        handleZipOnBlur,
        isCountriesLoading,
        states
    } = useShippingFields({
        getCountriesQuery: GET_COUNTRIES_QUERY,
        getStatesQuery: GET_STATES_QUERY,
        setIsFetchingMethods: setIsFetchingMethods,
        setShippingMutation: SET_SHIPPING_MUTATION
    });

    const { country, state, zip } = selectedShippingFields;
    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <Fragment>
            <Field
                id="country"
                label="Country"
                classes={{ root: classes.country }}
            >
                <Select
                    disabled={isCountriesLoading ? true : null}
                    field="country"
                    initialValue={country}
                    items={countries}
                />
            </Field>
            <Field id="state" label="State" classes={{ root: classes.state }}>
                <Select field="state" initialValue={state} items={states} />
            </Field>
            <Field id="zip" label="ZIP" classes={{ root: classes.zip }}>
                <TextInput
                    field="zip"
                    initialValue={zip}
                    onBlur={handleZipOnBlur}
                />
            </Field>
        </Fragment>
    );
};

export default ShippingFields;

ShippingFields.propTypes = {
    classes: shape({
        country: string,
        state: string,
        zip: string
    }),
    selectedShippingFields: shape({
        country: string.isRequired,
        state: string.isRequired,
        zip: string.isRequired
    }),
    setIsFetchingMethods: func
};

export const GET_COUNTRIES_QUERY = gql`
    query GetCountries {
        countries {
            full_name_english
            two_letter_abbreviation
        }
    }
`;

export const GET_STATES_QUERY = gql`
    query GetStates($countryCode: String!) {
        country(id: $countryCode) {
            id
            available_regions {
                id
                code
                name
            }
        }
    }
`;

export const SET_SHIPPING_MUTATION = gql`
    mutation SetShippingAddressForEstimate(
        $cartId: String!
        $address: CartAddressInput!
    ) {
        setShippingAddressesOnCart(
            input: {
                cart_id: $cartId
                shipping_addresses: [{ address: $address }]
            }
        ) {
            cart {
                ...ShippingMethodsFragment
            }
        }
    }
    ${ShippingMethodsFragment}
`;
