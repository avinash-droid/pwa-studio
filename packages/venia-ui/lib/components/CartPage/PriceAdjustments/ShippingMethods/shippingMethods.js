import React from 'react';
import gql from 'graphql-tag';
import { Form } from 'informed';

import { useShippingMethods } from '@magento/peregrine/lib/talons/CartPage/PriceAdjustments/ShippingMethods/useShippingMethods';

import { mergeClasses } from '../../../../classify';

import { ShippingMethodsFragment } from './shippingMethodsFragments';
import defaultClasses from './shippingMethods.css';
import ShippingRadios from './shippingRadios';
import ShippingFields from './shippingFields';

const ShippingMethods = props => {
    const { handleSubmit, shippingMethods } = useShippingMethods({
        getShippingMethodsQuery: GET_SHIPPING_METHODS
    });

    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <Form className={classes.root} onSubmit={handleSubmit}>
            <ShippingFields />
            <div className={classes.shipping_methods}>
                <h3 className={classes.prompt}>Shipping Methods</h3>
                <ShippingRadios shippingMethods={shippingMethods} />
            </div>
        </Form>
    );
};

export default ShippingMethods;

export const GET_SHIPPING_METHODS = gql`
    query GetShippingMethods($cartId: String!) {
        cart(cart_id: $cartId) {
            ...ShippingMethodsFragment
        }
    }
    ${ShippingMethodsFragment}
`;
