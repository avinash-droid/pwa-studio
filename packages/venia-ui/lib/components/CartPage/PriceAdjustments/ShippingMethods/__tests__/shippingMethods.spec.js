import React from 'react';
import { useLazyQuery } from '@apollo/react-hooks';
import { createTestInstance } from '@magento/peregrine';

import ShippingFields from '../shippingFields';
import ShippingMethods from '../shippingMethods';

jest.mock('../../../../../classify');

jest.mock('@apollo/react-hooks', () => {
    return { useLazyQuery: jest.fn() };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('../shippingFields', () => 'ShippingFields');
jest.mock('../shippingRadios', () => 'ShippingRadios');

test('renders loading message while fetching data', () => {
    useLazyQuery.mockReturnValueOnce([
        () => {},
        {
            called: true,
            loading: true
        }
    ]);

    const instance = createTestInstance(<ShippingMethods />);
    expect(instance.toJSON()).toMatchSnapshot();
});

test('renders form and shipping fields only when shipping address not entered', () => {
    useLazyQuery.mockReturnValueOnce([
        () => {},
        {
            called: true,
            loading: false,
            error: false,
            data: {
                cart: {
                    shipping_addresses: []
                }
            }
        }
    ]);

    const instance = createTestInstance(<ShippingMethods />);
    expect(instance.toJSON()).toMatchSnapshot();
});

test('renders shipping methods when shipping method is entered', () => {
    useLazyQuery.mockReturnValueOnce([
        () => {},
        {
            called: true,
            loading: false,
            error: false,
            data: {
                cart: {
                    shipping_addresses: [
                        {
                            available_shipping_methods: ['method1', 'method2'],
                            country: { code: 'US' },
                            postcode: '78758',
                            region: { code: 'TX' },
                            selected_shipping_method: {
                                carrier_code: 'usps',
                                method_code: 'method1'
                            }
                        }
                    ]
                }
            }
        }
    ]);

    const instance = createTestInstance(<ShippingMethods />);
    expect(instance.toJSON()).toMatchSnapshot();
});

test('applies mask to shipping methods while sibling fetch is in flight', () => {
    useLazyQuery.mockReturnValue([
        () => {},
        {
            called: true,
            loading: false,
            error: false,
            data: {
                cart: {
                    shipping_addresses: [
                        {
                            available_shipping_methods: ['method1', 'method2'],
                            country: { code: 'US' },
                            postcode: '78758',
                            region: { code: 'TX' }
                        }
                    ]
                }
            }
        }
    ]);

    const instance = createTestInstance(<ShippingMethods />);
    const { root } = instance;
    const { setIsFetchingMethods } = root.findByType(ShippingFields).props;
    setIsFetchingMethods(true);
    expect(instance.toJSON()).toMatchSnapshot();
});
