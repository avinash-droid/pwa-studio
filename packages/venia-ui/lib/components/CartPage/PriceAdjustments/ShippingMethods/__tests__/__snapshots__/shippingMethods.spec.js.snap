// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applies mask to shipping methods while sibling fetch is in flight 1`] = `
<form
  className="root"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <ShippingFields
    selectedShippingFields={
      Object {
        "country": "US",
        "state": "TX",
        "zip": "78758",
      }
    }
    setIsFetchingMethods={[Function]}
  />
  <div
    className="shipping_methods mask"
  >
    <h3
      className="prompt"
    >
      Shipping Methods
    </h3>
    <ShippingRadios
      selectedShippingMethod={null}
      shippingMethods={
        Array [
          "method1",
          "method2",
        ]
      }
    />
  </div>
</form>
`;

exports[`renders form and shipping fields only when shipping address not entered 1`] = `
<form
  className="root"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <ShippingFields
    selectedShippingFields={
      Object {
        "country": "US",
        "state": "",
        "zip": "",
      }
    }
    setIsFetchingMethods={[Function]}
  />
</form>
`;

exports[`renders loading message while fetching data 1`] = `
<span>
  Loading Shipping Data...
</span>
`;

exports[`renders shipping methods when shipping method is entered 1`] = `
<form
  className="root"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <ShippingFields
    selectedShippingFields={
      Object {
        "country": "US",
        "state": "TX",
        "zip": "78758",
      }
    }
    setIsFetchingMethods={[Function]}
  />
  <div
    className="shipping_methods "
  >
    <h3
      className="prompt"
    >
      Shipping Methods
    </h3>
    <ShippingRadios
      selectedShippingMethod="usps|method1"
      shippingMethods={
        Array [
          "method1",
          "method2",
        ]
      }
    />
  </div>
</form>
`;
